viskores/exec/CubicHermiteSpline.h

```diff
@@ -0,0 +1,145 @@
+                               const viskores::cont::ArrayHandle<viskores::Vec3f>& tangents,
+                               viskores::cont::DeviceAdapterId device,
+                               viskores::cont::Token& token)
+  {
+    this->Data = data.PrepareForInput(device, token);
+    this->Knots = knots.PrepareForInput(device, token);
+    this->Tangents = tangents.PrepareForInput(device, token);
+  }
+
+  VISKORES_EXEC viskores::ErrorCode Evaluate(const viskores::FloatDefault& tVal, viskores::Vec3f& val) const
+  {
+    viskores::Vec3f p0, p1, m0, m1;
+    viskores::FloatDefault t0, t1;
+
+    if (!this->GetInterval(tVal, p0, p1, t0, t1, m0, m1))
+      return viskores::ErrorCode::ValueOutOfRange;
+
+    val = this->EvaluatePosition(tVal, p0, p1, t0, t1, m0, m1);
+    return viskores::ErrorCode::Success;
+  }
+
+  private:
+
+  VISKORES_EXEC bool GetInterval(const viskores::FloatDefault& tVal,
+    viskores::Vec3f& p0, viskores::Vec3f& p1,
+    viskores::FloatDefault& t0, viskores::FloatDefault& t1,
+    viskores::Vec3f& m0, viskores::Vec3f& m1)const
+  {
+    viskores::Id idx = this->FindParametricInterval(tVal);
+    if (idx < 0)
+      return false;
+
+    t0 = this->Knots.Get(idx);
+    t1 = this->Knots.Get(idx + 1);
+    p0 = this->Data.Get(idx);
+    p1 = this->Data.Get(idx + 1);
+    m0 = this->Tangents.Get(idx);
+    m1 = this->Tangents.Get(idx + 1);
+
+    return true;
+  }
+
+  VISKORES_EXEC viskores::Vec3f EvaluatePosition(const viskores::FloatDefault& tVal,
+                                         const viskores::Vec3f& p0,
+                                         const viskores::Vec3f& p1,
+                                         const viskores::FloatDefault& t0,
+                                         const viskores::FloatDefault& t1,
+                                         const viskores::Vec3f& m0,
+                                         const viskores::Vec3f& m1) const
+  {
+    viskores::Vec3f result;
+
+    // Hermite basis functions.
+    viskores::FloatDefault dt = t1-t0;
+    viskores::FloatDefault tNorm = (tVal - t0) / dt;
+    viskores::FloatDefault t2 = tNorm * tNorm, t3 = t2 * tNorm;
+    viskores::FloatDefault h00 = 2 * t3 - 3 * t2 + 1;
+    viskores::FloatDefault h10 = t3 - 2 * t2 + tNorm;
+    viskores::FloatDefault h01 = -2 * t3 + 3 * t2;
+    viskores::FloatDefault h11 = t3 - t2;
+
+    for (viskores::Id i = 0; i < 3; ++i)
+      result[i] = h00 * p0[i] + h10 * dt * m0[i] + h01 * p1[i] + h11 * dt * m1[i];
+
+    return result;
+  }
+
+  viskores::Id FindParametricInterval(const viskores::FloatDefault& t) const
+  {
+    viskores::Id n = this->Knots.GetNumberOfValues();
+
+    if (t < this->Knots.Get(0) || t > this->Knots.Get(n - 1))
+      return -1;
+
+    //Binary search for the interval
+    viskores::Id left = 0;
+    viskores::Id right = n - 1;
+
+    while (left < right)
+    {
+      viskores::Id mid = left + (right - left) / 2;
```

Comment:
This is a weird way of saying

```suggestion
      viskores::Id mid = (right + left) / 2;
```